cpp/src/parquet/encoding_test.cc

```diff
@@ -540,8 +538,203 @@ void GetBinaryDictDecoder(DictEncoder<ByteArrayType>* encoder, int64_t num_value
 
   *out_values = buf;
   *out_dict = dict_buf;
-  *out_decoder = std::unique_ptr<ByteArrayDecoder>(
-      dynamic_cast<ByteArrayDecoder*>(decoder.release()));
+  ASSERT_NE(decoder, nullptr);
+  auto released = dynamic_cast<TypedDecoder<T>*>(decoder.release());
+  ASSERT_NE(released, nullptr);
+  *out_decoder = std::unique_ptr<TypedDecoder<T>>(released);
+}
+
+template <typename Pair = void>
+class EncodingAdHocTyped {
+ public:
+  using Types = ::testing::Types<std::pair<::arrow::BooleanType, BooleanType>,
+                                 std::pair<::arrow::Int32Type, Int32Type>,
+                                 std::pair<::arrow::Int64Type, Int64Type>,
+                                 std::pair<::arrow::FloatType, FloatType>,
+                                 std::pair<::arrow::DoubleType, DoubleType>,
+                                 std::pair<::arrow::FixedSizeBinaryType, FLBAType>>;
+};
```

Comment:
This idiom struck me as odd, as the contents does not vary by the passed type

Reply:
Not sure what you mean

Reply:
As far as I can tell this class doesn't "do" anything. `Types` does not depend on `Pair`

Reply:
This class is otherwise empty; I just use it as a container for the types pairs which actual test cases are instantiated: https://github.com/apache/arrow/pull/5441/files#diff-a1f922323295c1e3170c9f16eea148d7R723